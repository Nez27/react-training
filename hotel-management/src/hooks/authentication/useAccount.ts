import { useQuery } from '@tanstack/react-query';

// Services
import { getCurrentAccount } from '@service/authenticationService';

/**
 * Get data of account currently login
 * @returns The data of account currently login
 */
const useAccount = () => {
  let isAuthenticated: boolean = false;
  const { data: account, isPending } = useQuery({
    queryKey: ['account'],
    queryFn: getCurrentAccount,
  });

  if (account) {
    isAuthenticated = account.role === 'authenticated';
  }

  return { isPending, account, isAuthenticated };
};

export { useAccount };
