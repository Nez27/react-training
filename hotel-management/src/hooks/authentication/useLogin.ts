import { useMutation, useQueryClient } from '@tanstack/react-query';
import { useNavigate } from 'react-router-dom';
import toast from 'react-hot-toast';

// Services
import { login as loginFn } from '@service/authenticationService';

/**
 * Login web
 * @returns The login function and isPending boolean
 */
const useLogin = () => {
  const queryClient = useQueryClient();
  const navigate = useNavigate();

  const { mutate: login, isPending } = useMutation({
    mutationFn: loginFn,
    onSuccess: (account) => {
      queryClient.setQueryData(['account'], account.user);

      navigate('/');
    },
    onError: (err) => {
      console.error(err.message);
      toast.error('Email or password not correct!');
    },
  });

  return { login, isPending };
};

export { useLogin };
