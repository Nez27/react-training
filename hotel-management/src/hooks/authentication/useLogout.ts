import { useMutation, useQueryClient } from '@tanstack/react-query';
import { useNavigate } from 'react-router-dom';

// Services
import { logout as logoutFn } from '@service/authenticationService';

/**
 * Logout web
 * @returns The logout function and isPending boolean
 */
const useLogout = () => {
  const navigate = useNavigate();
  const queryClient = useQueryClient();

  const { mutate: logout, isPending } = useMutation({
    mutationFn: logoutFn,
    onSuccess: () => {
      queryClient.removeQueries();
      navigate('/login', { replace: true });
    },
  });

  return { logout, isPending };
};

export { useLogout };
