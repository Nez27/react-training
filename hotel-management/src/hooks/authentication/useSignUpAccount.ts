import { useMutation } from '@tanstack/react-query';
import { createAccount as createAccountFn } from '@src/services/authenticationService.ts';
import toast from 'react-hot-toast';

const useSignUpAccount = () => {
  const { mutate: createAccount, isPending: isCreating } = useMutation({
    mutationFn: createAccountFn,
    onSuccess: (account) => {
      toast.success(`Create user ${account!.user!.email} successful!`);
    },
    onError: (err) => toast.error(err.message),
  });

  return { createAccount, isCreating };
};

export { useSignUpAccount };
