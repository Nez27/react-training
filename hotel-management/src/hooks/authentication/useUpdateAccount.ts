import { useMutation, useQueryClient } from '@tanstack/react-query';
import { toast } from 'react-hot-toast';

// Services
import { updateAccount as updateAccountFn } from '@service/authenticationService';

// Constants
import { UPDATE_ACCOUNT_SUCCESSFUL } from '@constant/messages';

const useUpdateAccount = () => {
  const queryClient = useQueryClient();

  const { mutate: updateAccount, isPending: isUpdating } = useMutation({
    mutationFn: updateAccountFn,
    onSuccess: (account) => {
      toast.success(UPDATE_ACCOUNT_SUCCESSFUL);
      queryClient.setQueryData(['account'], account.user);
    },
    onError: (err) => toast.error(err.message),
  });

  return { updateAccount, isUpdating };
};

export { useUpdateAccount };
