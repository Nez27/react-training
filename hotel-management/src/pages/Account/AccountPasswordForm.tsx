import { useEffect } from 'react';
import { useForm } from 'react-hook-form';
import toast from 'react-hot-toast';

// Styled
import Input from '@src/commons/styles/Input';

// Components
import Form from '@src/components/Form';

// Constants
import { REGEX } from '@src/constants/commons';
import {
  PASSWORD_CONDITION,
  PASSWORD_NOT_MATCH,
  PASSWORD_NOT_MATCH_REQUIREMENT,
} from '@src/constants/formValidateMessage';

// Helpers
import { isValidRegex } from '@src/helpers/validators';

// Hooks
import { useUpdateAccount } from '@src/hooks/authentication/useUpdateAccount';

interface IAccountPasswordForm {
  password: string;
  confirmPassword: string;
}

const AccountPasswordForm = () => {
  const { updateAccount, isUpdating } = useUpdateAccount();
  const {
    register,
    handleSubmit,
    getValues,
    reset,
    formState: { errors },
  } = useForm<IAccountPasswordForm>();

  useEffect(() => {
    if (errors?.password?.message) {
      toast.error(PASSWORD_CONDITION);
    }
  }, [errors?.password?.message]);

  const onSubmit = (data: IAccountPasswordForm) => {
    updateAccount(data, {
      onSuccess: () => {
        reset();
      },
    });
  };

  return (
    <Form onSubmit={handleSubmit(onSubmit)}>
      <Form.Row label="Password:" error={errors?.password?.message}>
        <Input
          id="password"
          type="password"
          {...register('password', {
            validate: {
              checkPassword: (v) =>
                isValidRegex(REGEX.PASSWORD, v) ||
                PASSWORD_NOT_MATCH_REQUIREMENT,
            },
          })}
        />
      </Form.Row>

      <Form.Row
        label="Confirm Password:"
        error={errors?.confirmPassword?.message}
      >
        <Input
          id="confirmPassword"
          type="password"
          {...register('confirmPassword', {
            validate: (value) =>
              getValues().password === value || PASSWORD_NOT_MATCH,
          })}
        />
      </Form.Row>

      <Form.Action>
        <Form.Button type="submit" name="submit" disabled={isUpdating}>
          Update password
        </Form.Button>
      </Form.Action>
    </Form>
  );
};

export default AccountPasswordForm;
