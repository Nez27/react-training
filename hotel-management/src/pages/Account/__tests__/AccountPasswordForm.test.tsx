import {
  RenderResult,
  fireEvent,
  render,
  waitFor,
} from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

// Components
import AccountPasswordForm from '../AccountPasswordForm';

// Hooks
import { useUpdateAccount } from '@src/hooks/authentication/useUpdateAccount';

jest.mock('@src/hooks/authentication/useUpdateAccount');

describe('AccountPasswordForm', () => {
  let wrapper: RenderResult | null = null;
  const queryClient = new QueryClient();
  const mockUpdateAccount = jest.fn();

  beforeEach(() => {
    (useUpdateAccount as jest.Mock).mockReturnValue({
      updateAccount: mockUpdateAccount,
      isUpdating: false,
    });

    wrapper = render(
      <QueryClientProvider client={queryClient}>
        <AccountPasswordForm />
      </QueryClientProvider>
    );
  });

  test('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Should submit form successful', async () => {
    const passwordField = wrapper!.container.querySelector('#password');
    const comfirmPasswordField =
      wrapper!.container.querySelector('#confirmPassword');
    const submitButton = wrapper!.getByText('Update password');

    fireEvent.change(passwordField!, { target: { value: 'Phanhuuloi2@@1' } });
    fireEvent.change(comfirmPasswordField!, {
      target: { value: 'Phanhuuloi2@@1' },
    });
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(mockUpdateAccount).toHaveBeenCalled();
    });
  });

  test('Should display error message when input invalid field', async () => {
    const passwordField = wrapper!.container.querySelector('#password');
    const submitButton = wrapper!.getByText('Update password');

    fireEvent.change(passwordField!, { target: { value: '@' } });
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(wrapper!.getByText('Password not match!')).toBeInTheDocument();
    });
  });
});
