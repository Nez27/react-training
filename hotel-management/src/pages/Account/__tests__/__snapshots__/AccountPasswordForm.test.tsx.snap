// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountPasswordForm Should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="sc-dlniIP bCPNqw"
      >
        <div
          class="sc-eCAqax dgDilB"
          direction="horizontal"
        >
          <label
            class="sc-jSFipO cMYxin"
          >
            Password:
          </label>
          <div>
            <input
              class="sc-bdnyFh gXSWSo"
              id="password"
              name="password"
              type="password"
            />
          </div>
        </div>
        <div
          class="sc-eCAqax dgDilB"
          direction="horizontal"
        >
          <label
            class="sc-jSFipO cMYxin"
          >
            Confirm Password:
          </label>
          <div>
            <input
              class="sc-bdnyFh gXSWSo"
              id="confirmPassword"
              name="confirmPassword"
              type="password"
            />
          </div>
        </div>
        <div
          class="sc-hKFymg dteuGK"
        >
          <button
            class="sc-gtsqUy sc-iCoHzw iXmfC cwhbgM"
            name="submit"
            type="submit"
            variations="primary"
          >
            Update password
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="sc-dlniIP bCPNqw"
    >
      <div
        class="sc-eCAqax dgDilB"
        direction="horizontal"
      >
        <label
          class="sc-jSFipO cMYxin"
        >
          Password:
        </label>
        <div>
          <input
            class="sc-bdnyFh gXSWSo"
            id="password"
            name="password"
            type="password"
          />
        </div>
      </div>
      <div
        class="sc-eCAqax dgDilB"
        direction="horizontal"
      >
        <label
          class="sc-jSFipO cMYxin"
        >
          Confirm Password:
        </label>
        <div>
          <input
            class="sc-bdnyFh gXSWSo"
            id="confirmPassword"
            name="confirmPassword"
            type="password"
          />
        </div>
      </div>
      <div
        class="sc-hKFymg dteuGK"
      >
        <button
          class="sc-gtsqUy sc-iCoHzw iXmfC cwhbgM"
          name="submit"
          type="submit"
          variations="primary"
        >
          Update password
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
